internal/state/configuration.go

```diff
@@ -94,18 +96,54 @@ func buildConfiguration(graph *graph) Configuration {
 		return Configuration{}
 	}
 
-	upstreams := buildUpstreams(graph.Backends)
+	upstreams := buildUpstreams(graph.Gateway.Listeners)
 	httpServers, sslServers := buildServers(graph.Gateway.Listeners)
+	backendGroups := buildBackendGroups(graph.Gateway.Listeners)
 
 	config := Configuration{
-		HTTPServers: httpServers,
-		SSLServers:  sslServers,
-		Upstreams:   upstreams,
+		HTTPServers:   httpServers,
+		SSLServers:    sslServers,
+		Upstreams:     upstreams,
+		BackendGroups: backendGroups,
 	}
 
 	return config
 }
 
+func buildBackendGroups(listeners map[string]*listener) []BackendGroup {
+	// There can be duplicate backend groups if a route is attached to multiple listeners.
+	// We use a map to deduplicate them.
+	uniqueGroups := make(map[string]BackendGroup, 0)
+
+	for _, l := range listeners {
+
+		if !l.Valid {
+			continue
+		}
+
+		for _, r := range l.Routes {
+			for _, group := range r.BackendRefs.ByRule {
+				if _, ok := uniqueGroups[group.GroupName()]; !ok {
+					uniqueGroups[group.GroupName()] = group
+				}
+			}
+		}
+
+	}
+
+	groups := make([]BackendGroup, 0, len(uniqueGroups))
+	for _, group := range uniqueGroups {
+		groups = append(groups, group)
+	}
+
+	// sort upstreams for test-ability
+	sort.Slice(groups, func(i, j int) bool {
```

Comment:
How often does this function get called? Does any change require a rebuild of groups?

Does sorting here help anywhere outside of testing? Can this be done in tests if it's a no-op for the production runtime?

Reply:
Good callout. There's no function outside of the tests.

 We have this in a couple of different places in the code. I'll move the sorting into the tests.

Reply:
I moved the upstream & backend group sorting into the tests.